packages/vow/src/watch.js

```diff
@@ -74,58 +67,50 @@ const prepareWatcherKit = (zone, rejectionMeansRetry, watchVow) =>
      */
     (resolver, watcher) => {
       const state = {
-        vow: undefined,
+        vow: /** @type {unknown} */ (undefined),
         resolver,
         watcher,
       };
       return /** @type {Partial<typeof state>} */ (state);
     },
     {
-      vowSetter: {
-        /** @param {any} vow */
-        setVow(vow) {
-          this.state.vow = vow;
-        },
+      /** @type {Required<import('./watch-promise.js').PromiseWatcher>['onFulfilled']} */
+      onFulfilled(value) {
+        const { watcher, resolver } = this.state;
+        if (getVowPayload(value)) {
+          // We've been shortened, so reflect our state accordingly, and go again.
+          this.state.vow = value;
+          watchVow(value, this.self);
+          return undefined;
+        }
+        this.state.watcher = undefined;
+        this.state.resolver = undefined;
+        if (!resolver) {
+          return undefined;
+        } else if (watcher) {
+          settle(resolver, watcher, 'onFulfilled', value);
+        } else {
+          resolver.resolve(value);
+        }
       },
-      promiseWatcher: {
-        /** @type {Required<import('./watch-promise.js').PromiseWatcher>['onFulfilled']} */
-        onFulfilled(value) {
-          const { watcher, resolver } = this.state;
-          if (getVowPayload(value)) {
-            // We've been shortened, so reflect our state accordingly, and go again.
-            this.facets.vowSetter.setVow(value);
-            watchVow(value, this.facets.promiseWatcher);
-            return undefined;
-          }
-          this.state.watcher = undefined;
-          this.state.resolver = undefined;
-          if (!resolver) {
-            return undefined;
-          } else if (watcher) {
-            settle(resolver, watcher, 'onFulfilled', value);
-          } else {
-            resolver.resolve(value);
-          }
-        },
-        /** @type {Required<import('./watch-promise.js').PromiseWatcher>['onRejected']} */
-        onRejected(reason) {
-          const { watcher, resolver } = this.state;
-          if (rejectionMeansRetry(reason)) {
-            watchVow(this.state.vow, this.facets.promiseWatcher);
-            return;
-          }
-          this.state.resolver = undefined;
-          this.state.watcher = undefined;
-          if (!watcher) {
-            resolver && resolver.reject(reason);
-          } else if (!resolver) {
-            throw reason; // for host's unhandled rejection handler to catch
-          } else if (watcher.onRejected) {
-            settle(resolver, watcher, 'onRejected', reason);
-          } else {
-            resolver.reject(reason);
-          }
-        },
+      /** @type {Required<import('./watch-promise.js').PromiseWatcher>['onRejected']} */
+      onRejected(reason) {
+        const { vow, watcher, resolver } = this.state;
+        if (vow && rejectionMeansRetry(reason)) {
+          watchVow(vow, this.self);
```

Comment:
Which test covers this scenario, in which a watched promise resolves to a vow that is subsequently rejected with a reason recognized by `rejectionMeansRetry`, and what is this `watchVow` expected to accomplish in such a case? Because it seems to me like an infinite loop.

Reply:
These are the tests, in `agoric-sdk/packages/vow/test/test-disconnect.js`.
```
    ℹ testing (plan=disco,happy, watchVow=true)
    ℹ testing (plan=disco,sad, watchVow=true)
    ℹ testing (plan=disco,sad,disco,happy, watchVow=true)
    ℹ testing (plan=disco,disco,happy, watchVow=true)
    ℹ testing (plan=disco,disco,sad, watchVow=true)
```
It's not an infinite loop because `watchVow` obtains a fresh promise from the vow internals's `shorten()` method.

Reply:
I'll rename `watchVow` to `retryAndWatchVow`.

Reply:
Ah, I see. So it would only be infinite with a fake vow whose `shorten` method maliciously returned disconnection messages.